lib/internal/modules/esm/initialize_import_meta.js

```diff
@@ -1,39 +1,50 @@
 'use strict';
 
 const { getOptionValue } = require('internal/options');
-const experimentalImportMetaResolve =
-  getOptionValue('--experimental-import-meta-resolve');
-const {
-  PromisePrototypeThen,
-  PromiseReject,
-} = primordials;
-const asyncESM = require('internal/process/esm_loader');
-
-function createImportMetaResolve(defaultParentUrl) {
-  return async function resolve(specifier, parentUrl = defaultParentUrl) {
-    return PromisePrototypeThen(
-      asyncESM.esmLoader.resolve(specifier, parentUrl),
-      ({ url }) => url,
-      (error) => (
-        error.code === 'ERR_UNSUPPORTED_DIR_IMPORT' ?
-          error.url : PromiseReject(error))
-    );
+const experimentalImportMetaResolve = getOptionValue('--experimental-import-meta-resolve');
+
+/**
+ * Generate a function to be used as import.meta.resolve for a particular module.
+ * @param {string} defaultParentUrl The default base to use for resolution
+ * @param {typeof import('./loader.js').ModuleLoader} loader Reference to the current module loader
+ * @returns {(specifier: string, parentUrl?: string) => string} Function to assign to import.meta.resolve
+ */
+function createImportMetaResolve(defaultParentUrl, loader) {
+  return function resolve(specifier, parentUrl = defaultParentUrl) {
+    let url;
+
+    try {
+      ({ url } = loader.resolve(specifier, parentUrl));
+    } catch (error) {
```

Comment:
Is there a way to remove this `try/catch` block? Function blocks with try/catch can't be inlined in v8 afaik.

Reply:
AFAIK that was true a long time ago, but is no longer accurate. Did you test it yourself with e.g. `--trace-opt` or `--print-code` (I'm not an expert in optimising for V8, I'm not sure those are the most relevant flags btw).